### DISCUSSION TITLE
Supertab like recipe with blink.nvim

### DISCUSSION THREAD

**Discussion by madflow** [LazyVim/LazyVim#5152] [Q&A] [answered]
I had this recipe in my config for `nvim-cmp` :

https://www.lazyvim.org/configuration/recipes#supertab

This allowed me to "cycle" through the autocomplete suggestions with the `TAB` key.

Is there already a drop in replacement (or some config hints) for the new default completion plugin `blink.nvim`, to make this available?

**Comment by agriffis** [comment]
If you look at https://github.com/Saghen/blink.cmp

and click on "Default configuration"

then you can see this commented section:

```lua
  -- "super-tab" keymap
  --   you may want to set `completion.trigger.show_in_snippet = false`
  --   or use `completion.list.selection = "manual" | "auto_insert"`
  --
  --   ['<C-space>'] = { 'show', 'show_documentation', 'hide_documentation' },
  --   ['<C-e>'] = { 'hide', 'fallback' },
  --
  --   ['<Tab>'] = {
  --     function(cmp)
  --       if cmp.snippet_active() then return cmp.accept()
  --       else return cmp.select_and_accept() end
  --     end,
  --     'snippet_forward',
  --     'fallback'
  --   },
  --   ['<S-Tab>'] = { 'snippet_backward', 'fallback' },
  --
  --   ['<Up>'] = { 'select_prev', 'fallback' },
  --   ['<Down>'] = { 'select_next', 'fallback' },
  --   ['<C-p>'] = { 'select_prev', 'fallback' },
  --   ['<C-n>'] = { 'select_next', 'fallback' },
  --
  --   ['<C-b>'] = { 'scroll_documentation_up', 'fallback' },
  --   ['<C-f>'] = { 'scroll_documentation_down', 'fallback' },
```

so I think you could get a supertab-like mapping with this in your LazyVim plugins

```lua
{
  'Saghen/blink.cmp',
  opts = {
    keymap = { preset = 'super-tab' },
  },
}
```

The default LazyVim config also refers to the super-tab preset, so you might want to check how that affects anything for you. See https://github.com/LazyVim/LazyVim/blob/1599187175eca74da52e94d7d65a1e08bcedccd9/lua/lazyvim/plugins/extras/coding/blink.lua#L105-L119

**Reply by madflow** [reply]
@agriffis 

I should have added, that I already tried this. When I add this:

```lua
{
  'Saghen/blink.cmp',
  opts = {
    keymap = { preset = 'super-tab' },
  },
}
```

to my plugins configuration, this does not yield the result I am looking for. I cannot see that anything has changed, for that matter. Maybe something is interfering in my setup.

Does this work for you?

**Reply by tmlnv** [reply]
@madflow same with me. Moreover, I've checked a clean distribution in Docker, tab cycling does not work there also.

**Reply by madflow** [reply, marked as answer]
So while writing my reply - someone posted their solution:

https://github.com/Saghen/blink.cmp/discussions/576#discussioncomment-11596409

So this:

```lua
  {
    "Saghen/blink.cmp",
    opts = {
      keymap = {
        preset = "enter",
        ["<Tab>"] = { "select_next", "fallback" },
        ["<S-Tab>"] = { "select_prev", "fallback" },
      },
      -- maybe optional 
      completion = { list = { selection = "auto_insert" } },
    },
  },
```

Gives me the "cycling with tab". Something is not 100% like before, I think, I cannot really put my finger on it yet :D

**Reply by agriffis** [reply]
@madflow I'm sorry, I hadn't actually tried it, but I've been mucking around with blink so I shared what I knew (or thought I knew, at least)

I do think that the blink keymaps—along with LazyVim's usage of them—are still in flux. Hopefully things will settle down soon.

**Comment by b11e** [comment]
Looks like it's solved

here's a small doc pr that I started 
https://github.com/LazyVim/lazyvim.github.io/pull/130

**Reply by madflow** [reply]
@btctrl The posted solution is not the same, like the [supertab](https://github.com/ervandew/supertab) like experience I had with nvim-cmp. I just did not have the time to document what is different.  I can rewire my brain - I would prefer a drop-in replacement. Still figurring things out.

**Reply by madflow** [reply]
So for me 

```
 completion = { list = { selection = "auto_insert" } },
```

prevented the first item to be selected. Since this is usually the one I accept anyway - I had to press an extra tab. I removed this line and it works better now. I will close here.
